### PR TITLE
Ability to edit/change client context

### LINKED ISSUE #40: Ability to edit/change client context
Currently, the JEXL debugging tool lacks functionality for users to modify the client context dynamically during debugging sessions. This prevents users from simulating different client contexts. 

Users should have the ability to change the client context within the JEXL debugging tool. This feature should allow them to input different context data and see how these changes affect the evaluation of JEXL expressions.

### PR DESCRIPTION
**Pull request by RJAK11**
Add functionality to dynamically modify the client context within the JEXL debugging tool.

Fixes #40

### CODE REVIEW

File: src/ui/components/ExperimentBrowserPage.tsx [hunk 1]
```diff
@@ -103,7 +103,7 @@ const ExperimentBrowserPage: FC = () => {
           addToast({
             message: `Id successfully generated and copied to clipboard. Test Id: ${result}`,
             variant: "success",
-            autohide: true,
+            autohide: false,
```
Comment (brennie) [inline src/ui/components/ExperimentBrowserPage.tsx:106]: why is this change here?

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 2]
```diff
@@ -57,6 +76,47 @@ const JEXLDebuggerPage: FC = () => {
...
+        setClientContext((prevContext) => ({
+          ...prevContext,
+          [key]: value,
+        }));
+      } else {
+        setTempContext((prevContext) => ({
+          ...prevContext,
+          [key]: value,
+        }));
+        debouncedParseAndSetContext(key, value);
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:91]: This is defined below. Im surprised TSC isnt complaining, because this should be undefined.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 3]
```diff
@@ -57,6 +76,47 @@ const JEXLDebuggerPage: FC = () => {
...
+      }
+    },
+    [],
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:94]: This should be bound over `setClientContext`, `setTempContext` and `parseAndSetContext`

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 4]
```diff
@@ -57,6 +76,47 @@ const JEXLDebuggerPage: FC = () => {
...
+        });
+      }
+    },
+    [addToast],
+  );
+
+  const debouncedParseAndSetContext = useMemo(
+    () => debounce(parseAndSetContext, 1000),
+    [parseAndSetContext],
+  );
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:118]: We can skip the useCallback and put the fn directly into a useMemo:

```ts
const parseAndSetContext = useMemo(() => {
  return debounce((key: string, value: string) => {
    // ... actual fn impl here
  }, 1000);
}, [addToast, setClientContext]);
```

File: src/apis/nimbus.js [hunk 5]
```diff
@@ -126,11 +125,7 @@ var nimbus = class extends ExtensionAPI {
 
           async evaluateJEXL(expression, context = {}) {
             try {
-              const targetingContext = await new TargetingContext(context, {});
-
-              const result = await targetingContext.evalWithDefault(expression);
-
-              return result;
+              return await lazy.FilterExpressions.eval(expression, context);
```
Comment (brennie) [inline src/apis/nimbus.js:128]: Why this change?

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 6]
```diff
@@ -57,6 +76,47 @@ const JEXLDebuggerPage: FC = () => {
...
+          [key]: parsedValue,
+        }));
+      } catch (error) {
+        addToast({
+          message: `Error changing context: ${(error as Error).message ?? String(error)}`,
+          variant: "danger",
+        });
+      }
+    },
+    [addToast],
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:112]: should also be capturing setClientContext

**Review by brennie — CHANGES_REQUESTED** [review]

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 7]
```diff
@@ -109,15 +163,19 @@ const JEXLDebuggerPage: FC = () => {
                 {["number", "string", "boolean"].includes(typeof value) ? (
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:163]: because we overwrite the original context, we lose the original type information. We should keep the two contexts separate.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 8]
```diff
@@ -57,6 +76,41 @@ const JEXLDebuggerPage: FC = () => {
...
+          variant: "danger",
+        });
+      }
+    }, 1000);
+  }, [addToast, setClientContext]);
+
+  const handleContextChange = useCallback(
+    (key: string, value: string, isText: boolean) => {
+      if (isText) {
+        setClientContext((prevContext) => ({
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:99]: We should keep the context and user overrides separate, so that we keep type information of the original types (for rendering fields and parsing values from the formData).

We can then provide `{...clientContext, ...overrides}` to the eval function.

We would then also know if values had changed, so we could also add reset buttons for fields (optional)

File: src/apis/nimbus.js (hunk 5 above)
Comment (RJAK11) [reply src/apis/nimbus.js:128]: `targetingContext.evalWithDefault` seems to use the actual values of the user's os, locale, and a couple of other items rather than the values from the changed context that gets passed into TargetingContext and I found that `lazy.FilterExpressions.eval` only uses the values from the context passed into it so I made that change

File: src/ui/components/ExperimentBrowserPage.tsx (hunk 1 above)
Comment (RJAK11) [reply src/ui/components/ExperimentBrowserPage.tsx:106]: When I renamed the parameter from `permanent ` to `autohide`, I forgot to flip the boolean values since they are kind of opposites so I just made that change here.

File: src/ui/components/ContextField.tsx [hunk 9]
```diff
@@ -0,0 +1,90 @@
+import React from "react";
+import { Form } from "react-bootstrap";
+
+type ContextValue = object | string | boolean | number | Date;
```
Comment (brennie) [inline src/ui/components/ContextField.tsx:4]: This is defined in multiple files.

File: src/ui/components/ContextField.tsx [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+type ContextFieldProps = {
+  contextKey: string;
+  onChange: (key: string, value: ContextValue) => void;
+  contextValue: ContextValue;
+  fieldType: "number" | "string" | "boolean" | "date" | "object";
+};
+
+const ContextField: React.FC<ContextFieldProps> = ({
```
Comment (brennie) [inline src/ui/components/ContextField.tsx:13]: I would put this component inside the JEXLDebuggerPage.tsx file.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 11]
```diff
@@ -10,17 +10,42 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
 
 import { useToastsContext } from "../hooks/useToasts";
 import { evaluateJexl } from "../jexlParser";
+import ContextField from "./ContextField";
+
+type ContextValue = object | string | boolean | number | Date;
+
+function debounce<F extends (...args: any[]) => void>(
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:17]: this should go in something like `src/utils/functional.ts`

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 12]
```diff
@@ -10,17 +10,42 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+  };
+}
 
 const JEXLDebuggerPage: FC = () => {
-  const [clientContext, setClientContext] = useState({});
+  const [originalContext, setOriginalContext] = useState<
+    Record<string, ContextValue>
+  >({});
+  const [modifiedContext, setModifiedContext] = useState({});
+  const [formData, setFormData] = useState({});
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:35]: Please add types.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 13]
```diff
@@ -10,17 +10,42 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+    }, delay);
+  };
+}
 
 const JEXLDebuggerPage: FC = () => {
-  const [clientContext, setClientContext] = useState({});
+  const [originalContext, setOriginalContext] = useState<
+    Record<string, ContextValue>
+  >({});
+  const [modifiedContext, setModifiedContext] = useState({});
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:34]: Please add types.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 14]
```diff
@@ -10,17 +10,42 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+  const [formData, setFormData] = useState({});
   const [jexlExpression, setJexlExpression] = useState("");
   const [output, setOutput] = useState("");
   const { addToast } = useToastsContext();
 
   const fetchClientContext = useCallback(async () => {
+    setOriginalContext({});
+    setModifiedContext({});
+    setFormData({});
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:43]: Why these set calls?

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 15]
```diff
@@ -10,17 +10,42 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+
     try {
       const context = await browser.experiments.nimbus.getClientContext();
-      setClientContext(context);
+      setOriginalContext(context as Record<string, ContextValue>);
+      setModifiedContext(context);
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:48]: We don't need to do this. The modifiedContext only needs to contain overrides from the formData.

We should, however, serialize everything from context into formData at this point.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 16]
```diff
@@ -100,27 +205,26 @@ const JEXLDebuggerPage: FC = () => {
...
-                    readOnly
-                    value={JSON.stringify(value, null, 2)}
-                    className="p-3 w-50 grey-border long-text"
-                  />
-                )}
+                <ContextField
+                  contextKey={key}
+                  contextValue={
+                    (formData[key] ?? modifiedContext[key]) as ContextValue
+                  }
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:218]: if we populate the formData when we read the initial context, we can just pass formData[key] through here.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 17]
```diff
@@ -100,27 +205,26 @@ const JEXLDebuggerPage: FC = () => {
...
+                  onChange={(key, value) =>
+                    handleContextChange(
+                      key,
+                      String(value),
+                      getFieldType(originalContext[key]),
+                    )
+                  }
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:225]: This creates a new fn every time this component renders, which causes the child component to re-render, even if there is no change.

We're passing the fieldType though and the child component knows the key and the value, so it should be responsible for calling useCallback() to create a memoized callback.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 18]
```diff
@@ -55,13 +83,90 @@ const JEXLDebuggerPage: FC = () => {
...
+    },
+    [setModifiedContext, setFormData, parseAndSetContext],
   );
 
+  const getFieldType = (
+    value: ContextValue,
+  ): "number" | "string" | "boolean" | "object" | "date" => {
+    if (typeof value === "string" && !isNaN(Date.parse(value))) {
+      return "date";
+    }
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:157]: value should be a Date, shouldn't it? this should be `value instanceof Date`.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 19]
```diff
@@ -55,13 +83,90 @@ const JEXLDebuggerPage: FC = () => {
...
+        }
+      }
+    }, 1000);
+  }, [addToast, setModifiedContext]);
+
+  const handleContextChange = useCallback(
+    (
+      key: string,
+      value: string,
+      fieldType: "number" | "string" | "boolean" | "object" | "date",
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:123]: This type should be defined somewhere.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 20]
```diff
@@ -55,13 +83,90 @@ const JEXLDebuggerPage: FC = () => {
...
-  const memoizedClientContextEntries = useMemo(
-    () => Object.entries(clientContext),
-    [clientContext],
+  const parseAndSetContext = useMemo(() => {
+    return debounce((key: string, value: string, isNumber: boolean) => {
+      if (isNumber) {
+        if (!isNaN(parseInt(value))) {
+          setModifiedContext((prevContext) => ({
+            ...prevContext,
+            [key]: parseInt(value),
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:94]: we dont have to call parseInt twice.

File: src/ui/components/ContextField.tsx [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+  contextKey,
+  contextValue,
+  onChange,
+  fieldType,
+}) => {
+  const handleChange = (
```
Comment (brennie) [inline src/ui/components/ContextField.tsx:19]: use the useCallback hook to prevent re-renders of children.

**Review by brennie — CHANGES_REQUESTED** [review]

File: src/ui/components/JEXLDebuggerPage.tsx (hunk 15 above)
Comment (RJAK11) [reply src/ui/components/JEXLDebuggerPage.tsx]: Since we do 
```
const result = await evaluateJexl(jexlExpression, {
          ...originalContext,
          ...modifiedContext,
 });
```
for the JEXL Evaluation, I noticed that if we don't do `setModifiedContext(context);` after the refresh, then the evaluator outputs outdated information that doesn't match what the user sees in the context. Do you think I should still remove it and find a different workaround or did you have something else in mind?

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 22]
```diff
@@ -55,13 +160,93 @@ const JEXLDebuggerPage: FC = () => {
...
+        }
+      }
+    }, 1000);
+  }, [addToast, setModifiedContext]);
 
-  const memoizedClientContextEntries = useMemo(
-    () => Object.entries(clientContext),
-    [clientContext],
+  const handleContextChange = useCallback(
+    (key: string, value: string, fieldType: ContextValueString) => {
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:198]: Lets let `value: FormDataValue` (ie `string | bool`).

**Review by brennie — CHANGES_REQUESTED** [review]

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 23]
```diff
@@ -10,17 +10,119 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+  const handleChange = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
+      const newValue =
+        e.target.type === "checkbox"
+          ? (e.target as HTMLInputElement).checked
+          : e.target.value;
+      onChange(contextKey, newValue);
+    },
+    [contextKey, onChange],
+  );
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:40]: We can have a better typing story here.

onChange should take (key: string, value: string | boolean, fieldType: FieldType).

however, if fieldType === boolean, we always want value to be a boolean. Otherwise we want it to be a string.

We can accomplish this with overloads:

```ts
type OnChangeFn = {
    (key: string, value: boolean, fieldType: "boolean"): void;
    (key: string, value: string, fieldType: Exclude<FieldType, "boolean">): void;
};
```

We can also specialize the ContextField based on the type of value passed in:

```ts
type FieldType = "string" | "number" | "boolean" | "object" | "date";
type ContextFieldProps<TValue extends FormDataValue> = {
    // if TValue is boolean, this must be "boolean", otherwise its not "boolean"
    fieldType: TValue extends boolean ? "boolean": Exclude<FieldType, "boolean">;
    onChange: OnChangeFn;
    contextKey: string;
    value: FormDataValue;
};

function ContextField<TValue extends FormDataValue>({
  fieldType,
  onChange,
  contextKey,
  value,
}: ContextFieldProps<TValue>) {
// ...
}
```

You can see a full worked example at this playground [link](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQdMAnmFnAGLQgAiyMyANWQAbAK5sAvHFQwowAHYBzOAB84AIwgRhOefWasOXXvzhTsuaABMAPDLlKANEag8+gkeIB8+lm3bAWMJWACp+ZuT2CopkquSa2rqxamTyoiDqWFDJ5BDqAFZY1HFkJli0jOEA8vIAwgAWyEpY7PIRAN50cN1wABQA1lhMAFzSstHOAG6eWKMJOk3OBIHBYayjZPNJAJSjkxDAVvQ9fYMjYw6KUzOjUU5wy0GhfqMAogAeuGJWWDYBT2ssM5NloFvIyF5dnB9od6ABfeiVQy1CDyGBYd4wf7BAAKxDAqBsISEYjYGPR8is6E4rhMHlJXg6XR6j1WLzgxJmcHJWEp6C2TTgAH54qCkm9Pt9ftjnqxgQLwT5md1UQ0mopZnAamrmq1jj1LGjyQBpIa3cZKfXdaak0ac0nwxEEUTyfDAVFwFFGzEyokk8TczG8qkuNz8f1YLy9TrdVmyoHM1WNZqOZmG9GY01MVPWmapuGjL0ZrErKx4iAEv0zSFwGMG1EyODJqw6HUaiLYPAwAB0olQWFqImE6jw-V6vWVJ25o07+G7bawr0mvJgNgAEiEALIAGQAkvIwKIYK8dCAV3ENzuQuSAIKUZAnrBntFeSfbMyMutTuDAAh9OOAmYEhSCCiRNGQ75ft+PRJuqWC9OmJpDM4vRYN2-BQBq8BoHAl57geR6Ps+MDbN2uD1EUgxWEspaAtsVpTnC3LCP2taTtBcCwc0CGosWWbOGhGFYd2NriDRAJ+PR7E9HCk5ws4ADaXEas4iGZshDy0X4AC6UknJOlAwKIUBtL0NheDYAD0kLwgwdAYpAsBwD8BDIKIwiEC6boegAUq8AAa27cFg6iiIoGpQDiyAatGBacQURQkWxJyGo2CkEMY7jOP2WKZfw2kdlQPZ9lgADK-Don8eXIFGUEnGQqBkBsFHCMIEBkDm0FkDATVwLIYlyXp9byI2ymSJgRW9v2g6tSOuD9DY2rJhqrRRtJ3Q8d6MBZualzXLaFzRHECriWy6wcFprDvhIn7rXA8lwApOU0mGyDaW+iInIZxmmXd5l3ScNhFuSMpwGp21DBIDWxABfhQ3cMScXUy1YBI7RjUxomo+0GW0u4CnQ9pTHWQDPRA7xIOlmDFPqUwUM9ZpEmsFDCqxGNaMY9CMxo7jr0Ez1RNwCTHEnO0QtIDAFEPNobUAO5HRiyDgDoP7oAoXyiD8Vh9RA0j1BAst9VLBhklAxBQN2pPdGuBtYMuUDZerbC83SqvOUw8hK8AuDOLLbCy7xcBYX14RZOb6AUZQcCUIo7nIHIwhMJbIs9AAQkecB+5nwCtUHWDwJLbCZPIWDLDA6AQH+JtgxR81HQoRtsOROfa5Y4CoueCCWbJKfdOTW2g+DWZQwAjjDl2o6BYJs8jcEc7PzSY9z7RkGgVgEGQxOvhxVnbz02x0HCQA)

Comment (brennie) [reply src/ui/components/JEXLDebuggerPage.tsx:40]: (This is optional)

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 24]
```diff
@@ -10,17 +10,160 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+    },
+    [onChange, contextKey, fieldType],
+  );
+
+  switch (fieldType) {
+    case "number":
+      return (
+        <Form.Control
+          type="number"
+          value={String(value)}
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:70]: this is already a string, so you can do `value as string`.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 25]
```diff
@@ -10,17 +10,160 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+          onChange={handleChange}
+          className="p-3 w-50 grey-border short-text"
+        />
+      );
+
+    case "string":
+      return (
+        <Form.Control
+          type="text"
+          value={String(value)}
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:80]: this is already a string, so you can do `value as string`.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 26]
```diff
@@ -10,17 +10,160 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+          className="p-3 w-50 ms-4 ps-5 mb-2 large-checkbox short-text"
+        />
+      );
+
+    case "Date":
+      return (
+        <Form.Control
+          type="datetime-local"
+          step="1"
+          value={String(value)}
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:101]: this is already a string, so you can do `value as string`.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 27]
```diff
@@ -10,17 +10,160 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+          onChange={handleChange}
+          className="p-3 w-50 grey-border short-text"
+        />
+      );
+
+    case "object":
+      return (
+        <Form.Control
+          as="textarea"
+          value={String(value)}
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:111]: this is already a string, so you can do `value as string`.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 28]
```diff
@@ -10,17 +10,160 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+          onChange={handleChange}
+          className="p-3 w-50 grey-border short-text"
+        />
+      );
+
+    case "boolean":
+      return (
+        <Form.Check
+          type="checkbox"
+          checked={Boolean(value)}
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:90]: this is already a boolean, so you can do `value as boolean`.

File: src/ui/components/JEXLDebuggerPage.tsx [hunk 29]
```diff
@@ -10,17 +10,160 @@ import { Container, Row, Col, Form, Button } from "react-bootstrap";
...
+  const [formData, setFormData] = useState<Record<string, FormDataValue>>({});
   const [jexlExpression, setJexlExpression] = useState("");
   const [output, setOutput] = useState("");
   const { addToast } = useToastsContext();
 
   const fetchClientContext = useCallback(async () => {
     try {
       const context = await browser.experiments.nimbus.getClientContext();
-      setClientContext(context);
+      setOriginalContext(context as Record<string, ContextValue>);
```
Comment (brennie) [inline src/ui/components/JEXLDebuggerPage.tsx:138]: you can cast the context here, and then you dont need to cast it below on 144

```
const context = (await browser.experiments.nimbus.getClientContext()) as Record<string, ContextValue>;
```

**Review by brennie — APPROVED** [review]
r+ with the issues addressed
